TicKV: Implement a persistent key/value storage

alistair23 opened the pull request:
### Pull Request Overview

TicKV (Tiny Circular Key Value) is a small file system allowing key value pairs to be stored in Flash Memory.

Tock currently has a flash HIL, but there is no implementation on top of that HIL to reliably and persistently store data over a long period of time for multiple apps.

This is the first step in that direction by implementing a key/value filesystem designed to store key/value pairs in flash. This also addresses: https://github.com/tock/tock/issues/1692

The design has been implemented as a library, which hopefully will allow others to use it if they want to.

For details of the design read the PR changes, I'm not going to repeat the design and goals here.

#### Why write a new filesystem?

 * I couldn't find one already implemented in Rust
 * I couldn't find one already implemented that would meet Tock's strict requirements
 * I couldn't find one small and simple enough to just do what we need

#### Next steps

Step 1 is to get this merged into Tock.

Then I will have to connect it up with the OpenTitan flash controller and expose it to Tock.

At the same time I want to work on reliable app IDs that can be used for access control.

Finally I need to write a capsule that allows apps to read/write data to/from flash in a secure way in which we can enforce access controls.

### Testing Strategy

Unit tests.

### TODO or Help Wanted

Feedback mostly.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

bradjc commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,494 @@
+buffer obtained when reading flash. This could be reduced by instead performing
+multiple reads of a smaller fixed size.
+
+### Synchronous
+
+TickFS is fully synchronous and every call is blocking.
```
Do you have an idea as to how you would implement this? I'm guessing by using a new flash driver?

alistair23 replied:
I think no matter what a new flash driver would need to be implemented. The current Tock interfaces are very Tock specific and I was trying to make this generic.

There are two options (or both):
 1. Update TickFS to also support asynchronous operations, we can use the callback style method used in Tock. This would take some work, but isn't impossible. Then we would just need a small translation layer to convert between Tock and TickFS.
 1. Add a new flash interface using synchronous calls.

bradjc replied:
The flash HIL is read/write/erase page with callbacks. I don't see how that is overly Tock specific. There may have to be a shim layer to mostly do address conversion/restriction, but that would be a standard shim that any board with a flash driver could use.

The other benefit to using an asynchronous interface is that this FS would work on an external flash chip (like the nrf52840dk has) as well.

hudson-ayers replied:
It only requires unsafe code to construct a static buffer that is initially passed to the capsule using the filesystem, so it is at least contained within `main.rs`. But I see that if not for the requirement of async operations, an anonymous lifetime would work, which is much nicer as it would allow clients of the filesystem to use stack allocated buffers for filesystem reads/writes

phil-levis replied:
This needs to be asynchronous: we talked about this in a call. While on-chip flash can behave synchronously, that's only true if you're writing to the bank you are executing out of. If you have off-chip flash (e.g., over a SPI) you do not want to put a blocking interface on top of it.

alistair23 replied:
The interface in Tock will be async, see https://github.com/tock/tock/pull/2205 for how this is done.

bradjc commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,494 @@
+
+TickFS is fully synchronous and every call is blocking.
+
+This is unfortunately inherently different to Tock's asynchronous design. The
+justification for this design is that in general writing and erasing flash
+can be tricky if we are executing from the same bank. A synchronous design
+seems like an overall more useful implementation to allow this. It also means
+we block in the critical moments where a power loss would loose the data we
+want to preserve.
```
What does the complexity for executing from the same bank as the file system look like?

alistair23 replied:
That will be hardware dependent. Most flash controllers have limitations when writing to/from flash. For example you can't load other data from flash while performing an erase. In past experience usually interrupts are disabled during a flash erase to make it a critical section.

There is some discussion about this here: https://github.com/tock/tock/issues/1459#issuecomment-554742330

By making this a synchronous operation we can hide all of that complexity and just offload it to the `FlashController` implementation.

bradjc replied:
Making an asynchronous operation look synchronous is challenging, but Tock has support calls for deferred calls which enable a synchronous operation to look asynchronous. If a flash controller driver needs to block during an erase it can, then it can set a deferred call to trigger a callback. However, if the flash controller knows the operation can happen asynchronously, it still has to block when it could instead wait for an interrupt.

alistair23 replied:
That's a good idea. The defered calls should let us re-use Tocks' Flash HIL with this current interface.

hudson-ayers commented:
> Tock currently has a flash HIL, but there is no implementation on top of that HIL to reliably and persistently store data over a long period of time for multiple apps.

This is true, but it is worth noting that the log interface exists (https://github.com/tock/tock/blob/master/capsules/src/log.rs), which allows for persistent storage of values in an append-only log across restarts. The log interface is missing some of the functionality in this PR - keys are just sequentially increasing numbers based on the number of writes, rather than allowing arbitrary user-set keys, and entries cannot be modified, only new entries can be added. Also, the log is not exposed to applications yet, and is not yet virtualized within the kernel.

hudson-ayers commented on libraries/tickfs/src/tickfs.rs:
```diff
@@ -0,0 +1,673 @@
+    }
+
+    /// Generate the hash and region number from a key
+    fn get_hash_and_region(&self, hash_function: &mut H, key: &str) -> (u64, usize) {
```
The discussion in https://github.com/tock/tock/issues/1692 seemed to conclude that having keys be `&[u8]` or `u32` values, instead of `&str` values, would be preferred. Can you elaborate on why you chose `&str` keys?

alistair23 replied:
I have changed it to `&[u8]`

alistair23 commented:
Ping!

phil-levis commented on libraries/tickfs/README.md:
```diff
@@ -0,0 +1,95 @@
+### Collisions
+
+TickFS will prevent a new key/value pair with a colliding hash of the key to be
+added. The collision will be reported to the user with the `KeyAlreadyExists`
```
Should this be `ErrorCode::KeyAlreadyExists`?

phil-levis commented on libraries/tickfs/README.md:
```diff
@@ -0,0 +1,95 @@
+
+### Power loss protection
+
+TickFS ensures that in the event of a power loss, no stored data is lost or
```
Can you phrase TickFS's guarantees in terms of ACID semantics? "No stored data is lost or corrupted" could mean a lot of things. E.g., what if the data isn't stored? I think what you're getting at is both Atomicity and Durability but I am not sure.

alistair23 replied:
Done.

phil-levis commented on libraries/tickfs/README.md:
```diff
@@ -0,0 +1,95 @@
+
+### Security
+
+TickFS uses check sums to check data integrity. TickFS does not have any measures
```
Checksums, CRCs, or some form of hash? Or undefined/implementation-specific? Why are checksums needed?

alistair23 replied:
Checksums are used for data integrity to ensure bits don't flip on the flash. There are more details in the spec document.

phil-levis commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,494 @@
+#### Check Sum
+
+The check sum is a hash of the entire object (not including the check sum).
+The check sum is calculated using the same hash algorithm used for the keys.
```
Then it's not a checksum.

Please note that hashes (especially cryptographic hashes) and CRCs have very different error detection behaviors. A CRC (depending on the generator polynomial) provides guarantees on certain errors it can detect (e.g., an odd number of bit errors). A hash provides no guarantees. A strong cryptographic hash, for example, means that flipping one bit in the input makes every bit in the output effectively random. This means with N bits there is a 2^-N probability you get a collision. It can be OK to make this tradeoff, but a CRC is actually in some ways better.

alistair23 replied:
Yep, I thought about this.

I would prefer to use a CRC here, the problem is we then also need a CRC implementation as well as a hash implementation in the kernel. The hope is that the hash implementation can be offloaded to hardware, so re-using it for a checksum will hopefully speed things up.

I think I'll move toward using a CRC though, as I agree it will provide much more useful error checking.

phil-levis commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,494 @@
+`core::hash::Hasher` trait.
+
+For the hash of the key a 64-bit value can be justified by the lack of
+collision avoidance in the implementation. If two keys have the same
```
Could you add salt, to avoid collisions? E.g., if you insert and it fails due to a collision, you can retry with a different salt?

alistair23 replied:
That's an option, but how do we find the salt when looking up the keys? Do we just have a list of salts to walk through?

phil-levis commented on libraries/tickfs/src/flash_controller.rs:
```diff
@@ -0,0 +1,79 @@
+//! The Flash Controller interface with hardware
```
What is the relationship between a FlashController and an allocated in-kernel storage region made with the `storage_volume` macro? Does this represent a controller over a logical region (e.g., a storage region) or the physical entire flash? It seems like TickFS should operate over a storage region -- is that how it's intended to operate?

phil-levis reviewed (commented):
Overall, I think this looks like a good design. It's nice that it's lean and fits in 673 lines of code. 

It's a little hard to do a detailed code review for bugs because the guarantees of operations are not 100% clear. I think it would be good to phrase the specification using TRD terms, so saying what an implementation MUST do, or SHOULD do, or MAY do. Separating its layout and API from a specific implementation would be good.

bradjc commented on libraries/tickfs/src/flash_controller.rs:
```diff
@@ -0,0 +1,79 @@
+///
+/// This is the public trait for the Flash controller implementation.
+///
+/// The PageSize 'P' must be the smallest size that can be erased in
```
Does `PageSize` refer to something in the code?

alistair23 replied:
Good catch, this is an out of date comment. I'll remove it.

bradjc commented on libraries/tickfs/src/flash_controller.rs:
```diff
@@ -0,0 +1,79 @@
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::WriteFail.
+    fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode>;
```
This doesn't need a region num?

alistair23 replied:
No, it's much more efficient to just write the smallest amount of data possible to a specific address. Generally flash controllers don't require an entire page write.

bradjc replied:
Maybe this should just be called `write()` then?

alistair23 replied:
Good point, I'll fix this.

bradjc commented on libraries/tickfs/src/flash_controller.rs:
```diff
@@ -0,0 +1,79 @@
+/// a single operation.
+/// The start and end address of the FlashController must be aligned
+/// to the PageSize.
+/// All `region_number`s and `address`es are offset from zero. If you
```
How do an implementer and user agree on how large a region is?

alistair23 replied:
I have updated the documentation to make this clearer.

bradjc replied:
What about making `get_region_size()` a member function in the trait (or maybe some other rust-y approach)?

alistair23 replied:
With const generic arrays I think this would be a little clearer. I'll see if I can find a rust-y way to make this nicer.

bradjc commented on libraries/tickfs/src/flash_controller.rs:
````diff
@@ -0,0 +1,81 @@
+/// }
+/// ```
+pub trait FlashController {
+    /// This function must read the data from the flash region specified by
+    /// `region_number` into `buf`. The length of the data read should be the
+    /// same length as buf. `offset` indicates an offset into the region that
+    /// should be read.
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::ReadFail.
+    fn read_region(
+        &self,
+        region_number: usize,
+        offset: usize,
+        buf: &mut [u8],
+    ) -> Result<(), ErrorCode>;
````
Is `buf` required to end up entirely within a single region?

alistair23 replied:
So right now buf will contain the entire region size. In future it might be worth allowing a smaller buf and read part way through a region.

alistair23 commented:
Ok, I have pushed an update.

@hudson-ayers and @bradjc I have included the async changes in this PR now, hopefully that will help with review. I have also included a little more justification for the API at the end of the SPEC document. They basically say the same thing I was saying this morning, in that I think the sync design is helpful for others and unit testing and that due to the loop nature of lookups even with callbacks we will still end up with a retry style loop anyway after the callback has occurred.

@bradjc I have switched over to const generics for the flash region accesses. This then nicely ensures that the buffers all match the region sizes and that it is clear to users how everything should work together. If we don't want to introduce an unstable feature I can revert the change, I feel it indicates what it will look like eventually anyway.

Let me know what you think.

bradjc commented on libraries/tickfs/src/tickfs.rs:
```diff
@@ -0,0 +1,930 @@
+                    RubbishState::ReadRegion(reg) => reg as isize,
+                    _ => unreachable!(),
+                },
+                _ => unreachable!(),
```
What prevents a user from calling `read_key()`, then an async operation starting, but then calling `append_key()` before calling `continue_op`? I.e., why can't the current state be `GetRead`?

alistair23 replied:
This prevents that. It doesn't support two concurrent operations.

Handling that would be very complex. What if two append_key operations called, how do we then make guarantees about ordering and not loosing data?

Only 1 operation can be in progress at a time, we require the operation to complete before a new one can start.

phil-levis commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+ 1. Erase the above region
+ 1. Move the valid objects from the reserved region to the original region,
+    after being defraged in memory
+ 1. Erase the reserved region
```
Does this mean that the reserved region is erased for every erase? I.e. it will fail first.

alistair23 replied:
This is just an example. I would expect that a reserved region would end up being used for every erase yes. It would be possible to move the reserved region (it changes every time to the last region you wanted to erase) but that then makes lookups slower as we loose ordering.

phil-levis commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+
+As for the check-sum. It was seriously considered to change this to a
+32-bit or even 16-bit CRC. This would reduce the overhead for each object
+by 4 or 6 bytes. Although a CRC is relatively easy to implement, it can still
```
How long is it in comparison to a flash erase/write? This seems like a handwavy "well I'd have to implement it". 

What are typical error patterns in MCU flash chips? It would be good to see what other systems do and why.

alistair23 replied:
Good questions. I honestly don't know.

I was originally torn what to do here, I was starting to think a CRC would be better here. The reason I didn't is that I know some MCUs offer CRC offloading, in which case I'm assuming it is slow enough that a hardware off loader becomes beneficial.

phil-levis replied:
I think you want to do some back of the envelope calcuations on how long it takes to compute a reasonable CRC and hash compared to the time it takes to erase a page and time it takes to write a page. I think you will find they are quite fast.

phil-levis commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+ensure correctness in the design. This is more difficult and complex to do in
+a async only operation.
+
+The async functions (see the documentation) can be used to implement a fully
```
If you want TickFS to be completely general for the type of NV storage it's on top of, having async and sync APIs makes sense. For example, FRAM is close enough to RAM speeds (and is usually integrated with RAM e.g. in terms of addressing) that it's not useful to have an async API.

Can this document describe the API and how this works?

Why can't the callback that tells a client that an async operation has completed be part of the API? It could be a client doesn't bother registering these callbacks, or a synchronous implementation doesn't ever issue them. I think it's very important that this callback originate from the same module, at least -- you shouldn't have to register a callback with some board-and-TickFS-implementation-specific callback to be able to use TickFS.

Do you retry an operation because it wasn't complete? Or because the implementation was busy?

alistair23 replied:
This is the spec documetnation, it describes the underlying spec and not the usage interface. The information for how to use the API is in the rust docs.

Do you mean add a Tock callback function to the API that is then set as a client in the kernel?

We retry because it wasn't complete.

phil-levis commented on libraries/tickfs/README.md:
```diff
@@ -0,0 +1,93 @@
+TickFS is based on similar concepts as
+[Yaffs1](https://yaffs.net/documents/how-yaffs-works]).
+
+TickFS provides ACID properties.
```
How does it provide isolation? This implies a particular ordering of different clients. It would be good to explain how it provides each.

alistair23 replied:
It doesn't support concurrency so everything is isolated :)

I'll add some explanation and probably remove isolation.

phil-levis replied:
ACID semantics are usually defined in terms of transactions, so you need to state what a transaction is. A single read or write? If you define its transactions as single operations, and there are no concurrent operations, then yes, it has isolation. But this means, for example, that a client can't safely check for a key, and if it's not present then insert it.

alistair23 commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+This allows us to quickly determine which flash region a key is stored in. This
+should improve lookup time by reducing the number of reads required.
+
+If a give region is full, we will start to look in neighboring regions. We first
```
Thanks for pointing those out. Doesn't double hashing have the disadvantage of having to calculate two hashes? That could potentially slow down lookups. Also it seems to take up more space right?

Cuckoo hashing changes the medium right? Which is difficult to do safely (power loss resilient) in flash.

I'm all on board with other methods, it is just everything seems to have compromises and it's hard to get the balance right.

phil-levis replied:
Depending on the second hash, you could be orders of faster on speed. The time it takes to parse a single header is probably greater than to compute a simple hash. Note that the second hash can be much simpler, if the first one is strong. E.g., you could divide by the number of regions and then mod by the number of regions (assuming you don't have too many regions).

alistair23 commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+
+### TickFS Regions
+
+A TickFS region is the smallest set of contiguous flash storage space that can
```
Good questions.

Right now it can not be larger. The main reason is to avoid having to do multiple reads once we have found a key. Also writing over flash bounds will be a little more tricky. I think this is something that could be added in the future. It's not a fundamental limitation, just an implementation limitation.

There is a KV header entry. Right now it's empty. I will look at saving the region size in there.

phil-levis replied:
You mention multi-page writes: what are the durability and atomicity guarantees of flash writes/erases? What happens if you lose power in the middle of a write/erase? If you do not have atomicity guarantees here then you have to design the system differently to provide ACID semantics.

alistair23 replied:
Aren't you always going to have that problem though? If we loss power when some data has been written, but not all the data then we are going to loose that key. Is there much we can do about that?

phil-levis replied:
You do not always have that problem: https://en.wikipedia.org/wiki/Write-ahead_logging

This is how all modern file systems and databases are built. Filesystems usually rely on block atomicity, which storage units provide (i.e., have enough capacitance to complete an operation once started). But they use this block atomicity to build higher-level atomicity. Note that in RAM-scarce settings, WAL does require two writes.

It's OK if TicKV doesn't provide full ACID. But you should be very clear about what guarantees it does provide.

alistair23 replied:
Ah, yep. WAL doesn't really work on flash though, at least not without wear levelling issues.

Ok, I think I have described that. We are dependent on the flash implementation for the operations. Let me know if you think I should change the README.

alistair23 commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+We continue this loop until we either find the key we are looking for or
+find a version 0xFF, indicating the end of the blocks in that region.
+
+This method allows a quick retrieval of data from a given key. We only
```
When all regions are full then yes. A lookup will search all of flash.

I can't really think of a way to avoid doing that though. Having some sort of list of keys would use space, and I think space is going to be too important for that.

alistair23 commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+
+### Synchronous and Asynchronous Usage
+
+TickFS supports both an synchronous and asynchronous usage. For developers
```
This is the spec document. The idea is that it covers the spec in details for those interested. Most users won't need to read this, instead they can read the Rust documentation (auto generated from the code) to see the API. This also ensures that cargo will test the documentation to ensure it's always up to date.

alistair23 commented on libraries/tickfs/SPEC.md:
```diff
@@ -0,0 +1,509 @@
+only a single flash bank all flash writes and erases must be done synchronously.
+The synchronous API should allow these type of operations to occur.
+
+On top of that it is much simpler to unit test a synchronous API. This means the
```
I would disagree. Part of designing something is to allow it to be easy to test. TickFS would be useless if we can't reliably test it for both successes and failures.

phil-levis replied:
But the point isn't whether you can test it, but rather how much work writing those tests are. Of course you can test an asynchronous system.

alistair23 commented:
@bradjc, @hudson-ayers, @lschuermann I have re-worked the async interface. It no longer uses retries and is more in line with Tock style code.

This has the disadvantage of now requiring static lifetime buffers. Which requires unsafe to create and in the Tock kernel we need to copy data from app slices to kernel buffers, which now means we need even more static buffers.

Let me know what you think.

alistair23 commented:
@phil-levis I updated some parts of the spec and readme and also changed the name.

Pending spec issues:
 - What should happen to the checksum? Change to a CRC, remove the checksum?
 - Options to avoid collisions. Use a salt?
 - Possibly swap to double hash or cuckoo hashing

Am I missing anything?

phil-levis commented on libraries/tickv/README.md:
```diff
@@ -0,0 +1,110 @@
+Atomicity: TicKV does not provide atomicity. If a operation fails mid write
+then there will be an invalid key/value stored to flash.
+
+Consistency: Consistency is maintained similar to atomicity. All operations
```
What does "similar to atomicity" mean? The atomicity comment says a database can have an invalid key/value. So it is possible for an operation to take the store into an invalid state?

It is useful to state when it *is* atomic or consistent. E.g., you can say that "Under the constraint that underlying flash operations succeed, TicKV provides atomicity. If the underlying flash has an error, the system crashes mid-operation, or power is lost mid-operation, an operation may be partially committed."

phil-levis replied:
What happens if a write fails? Is the rest of the page considered full?

alistair23 replied:
I have updated this in the doc to answer all your questions. Let me know what you think. I have pasted it below as well:

> ## System Guarantees
> 
> A successful write or erase is durably committed.
> 
> The design does not support concurrency, such that it imposes a total order
> on all read, write and delete operations. Successful individual operations
> are therefore atomic. Applications that require higher-level atomicity
> (e.g., read/modify/delete/write, multiple writes) need to build this on top
> of these operations.
> 
> TicKV is not robust to low-level flash failures, power loss, or system
> crashes. However, a failure only affects a single key: a failure to write
> or delete key K will corrupt at most only K. A value to read will have no
> effect.
> 
> Under the constraint that underlying flash operations succeed, TicKV provides
> atomicity. If the underlying flash has an error, the system crashes
> mid-operation, or power is lost mid-operation, an operation may be
> partially committed.
> 
> If a write fails with only a partial write to flash the entire length of
> the key will be used in the storage.
> 
> TicKV ensures durability and once a transaction has completed
> and been committed to flash it will remain there. TicKV also takes measures
> to apply wear levelling to the flash storage.

phil-levis commented on libraries/tickv/README.md:
```diff
@@ -0,0 +1,110 @@
+provides isolation. The layer above TicKV is responsible for handling
+concurrent accesses by deferring operations for example.
+
+Durability: TicKV ensures durability and once a transaction has completed
```
Should put a note about wear-leveling here.

phil-levis commented on libraries/tickv/README.md:
```diff
@@ -0,0 +1,110 @@
+ * Keys will rarely be deleted
+ * Key values will rarely need to be modified
+
+## ACID characteristics
```
I would just call this "System Guarantees"

phil-levis commented on libraries/tickv/SPEC.md:
```diff
@@ -0,0 +1,510 @@
+remove anything from flash. Instead calling `invalidate_key()`
+will mark the `valid` boolean for that object as `false` (0).
+
+This is done because changing a `1` to a `0` in flash can be done with a
```
Many chips allow a limited number of writes per erase cycle on a page. E.g., the EFMPG1 Reference Manual reads:

"It is possible to write words twice between each erase by keeping at 1 the bits that are not to be changed. Let us take as an example
writing two 16 bit values, 0xAAAA and 0x5555. To safely write them in the same flash word this method can be used:
• Write 0xFFFFAAAA (word in flash becomes 0xFFFFAAAA)
• Write 0x5555FFFF (word in flash becomes 0x5555AAAA)
Note:
• There is a maximum of two writes to the same word between each erase due to a physical limitation of the flash.
• During a write or erase, flash read accesses will be stalled, effectively halting code execution from flash. Code execution continues
upon write/erase completion. Code residing in RAM may be executed during a write/erase operation."

It's OK to say that TicKV only works on flash devices that allow N>K writes to a word. It looks like TicKV requires 2 writes be allowed. That means almost all (but perhaps not all) can work. You want this limitation up at the top, not at the bottom.

What happens if you re-insert a key? Does it fail or do you replace?

alistair23 replied:
Yeah, I have seen some devices with limitations.

As long as we can write twice between erases. Once to store the data and once to invalidate then it should work. From what I can tell it's not very common that you can't write that many times.

I have added a note to the README.

To re-insert a key you first have to delete the key then append it again. It won't replace the original key, it will be appended. The first key (that was invalidated) is marked for deletion and can't be read.

phil-levis reviewed (commented):
The absence of any real in-memory data structures is a limitation here. It means, for example, once a key write is corrupted there's no way out, because you can't do a page-remapping. It could be, for example, that each page says its logical page number (which might be different than its physical), and at boot software scans the pages and build a logical->physical table mapping.

Given how far the design is from ACID, you might just want to instead explicitly state the guarantees:
  - A successful write or erase is durably committed.
  - The design does not support concurrency, such that it imposes a total order on all read, write, and delete. operations. Successful individual operations are therefore atomic. Applications that require higher-level atomicity (e.g., read/modify/delete/write, multiple writes) need to build this on top of these operations.
  -  TicKV does is not robust to low-level flash failures, power loss, or system crashes. However, a failure only affects a single key: a failure to write or delete key K will corrupt at most only K (*is this correct? I think it is from the append structure*).

bradjc reviewed (commented):
note: I'm not reviewing the library itself.

At the risk of going in circles, I think `async_ops.rs` can be removed from this PR. I believe that file can be entirely contained in a Tock capsule which implements the future KV HIL using TickV. Otherwise I think this library is a good addition to Tock (and perhaps should be spun out someday!).

alistair23 commented on libraries/tickv/src/async_ops.rs:
```diff
@@ -0,0 +1,923 @@
+
+/// The struct storing all of the TicKV information for the async implementation.
+pub struct AsyncTicKV<'a, C: FlashController<S>, H: Hasher, const S: usize> {
+    /// The controller used for flash commands
```
I agree that it can be removed with the HIL. Whatever order it all ends up being approved I can rebase from there.

There will be more changes required to match the new HIL, but that should be easy to address.

alistair23 commented:
> note: I'm not reviewing the library itself.
> 
> At the risk of going in circles, I think `async_ops.rs` can be removed from this PR. I believe that file can be entirely contained in a Tock capsule which implements the future KV HIL using TickV. Otherwise I think this library is a good addition to Tock (and perhaps should be spun out someday!).

I agree that it can be removed with the HIL. Whatever order it all ends up being approved I can rebase from there.

There will be more changes required to match the new HIL, but that should be easy to address.

alistair23 commented:
> The absence of any real in-memory data structures is a limitation here. It means, for example, once a key write is corrupted there's no way out, because you can't do a page-remapping. It could be, for example, that each page says its logical page number (which might be different than its physical), and at boot software scans the pages and build a logical->physical table mapping.

Yes, it will be saved in flash. The implementation can probably be improved to allow the invalid key to be marked as invalid, so that it won't conflict if a new key is added. That way if a corrupt key is saved a real key can be saved again on reboot.

The logical->physical map is a good idea. I think that should be possible to add in the future, without a major re-work. It will have the disadvantage of larger memory usage, but it might be worth it.

> 
> Given how far the design is from ACID, you should say it's not a ACID system and instead explicitly state the guarantees:
> 
>     * A successful write or erase is durably committed.
> 
>     * The design does not support concurrency, such that it imposes a total order on all read, write, and delete. operations. Successful individual operations are therefore atomic. Applications that require higher-level atomicity (e.g., read/modify/delete/write, multiple writes) need to build this on top of these operations.
> 
>     * TicKV does is not robust to low-level flash failures, power loss, or system crashes. However, a failure only affects a single key: a failure to write or delete key K will corrupt at most only K (_is this correct? I think it is from the append structure_).

I put most of this in the README.

bradjc commented on libraries/tickv/src/lib.rs:
```diff
@@ -0,0 +1,241 @@
+//!      any time.
+//!
+
+#![feature(min_const_generics)]
+#![no_std]
+#![deny(missing_docs)]
```
```suggestion
#![feature(min_const_generics)]
#![no_std]
#![forbid(unsafe_code)]
#![deny(missing_docs)]
```

hudson-ayers replied:
once rebased, `min_const_generics` is no longer needed

bradjc reviewed (approved):
Alright in my opinion this is great addition to Tock. It's an sync and async k-v implementation entirely in safe rust.

We also now have a low-level key-value store HIL interface which I believe a small shim capsule should be able to provide that trait with this implementation. Then ~~we~~ K-V master Alistair can build the rest of the stack.

@phil-levis any remaining issues?

bradjc commented on libraries/tickv/src/lib.rs:
```diff
@@ -0,0 +1,241 @@
+//!      any time.
+//!
+
+#![feature(min_const_generics)]
```
```suggestion
```

alistair23 commented:
Rebased and addresses comments

bradjc commented:
bors r+
